### PR TITLE
Add subtests to test.Assert

### PR DESCRIPTION
**Pull request by ivokub**
Added method `Run(fn func(assert *test.Assert), descs ...string)` to Assert type. This is similar to the `Run` method in Go's standard testing package to allow running subtests. Refactored tests in places to use subtests.

Removed explicit `Fuzz` call in integration tests as Fuzz is already called in `Assert.ProverSucceeded`.

An potential added benefit of using subtests is better coverage with race detector and enabling running subtests in parallel. I didn't implement it now though as Assert uses unsynchronized cache for compiled circuits.

### CODE REVIEW

File: circuitstats_test.go [hunk 1]
```diff
@@ -3,41 +3,40 @@ package gnark
...
-				nbConstraints := ccs.GetNbConstraints()
-				internal, secret, public := ccs.GetNbVariables()
-				checkStats(t, name, nbConstraints, internal, secret, public, curve, backendID)
+func TestCircuitStatistics(t *testing.T) {
+	assert := test.NewAssert(t)
+	for name := range circuits.Circuits {
+		for _, curve := range ecc.Implemented() {
+			for _, b := range backend.Implemented() {
+				// copy the circuit now in case assert calls t.Parallel()
+				tData := circuits.Circuits[name]
```
Comment (gbotrel) [inline circuitstats_test.go:29]: --> Circuits store `frontend.Circuit` which is an interface, this is not copying the actual underlying value. 

Here I would probably just go for a lock in the internal/backend/TestCircuit data structure.

**Review by gbotrel — CHANGES_REQUESTED** [review]

**Comment by ivokub** [comment]
I went over again over the different range loops and initialized the iteration-local variables everywhere they were used locally.

File: circuitstats_test.go (hunk 1 above)
Comment (gbotrel) [reply circuitstats_test.go]: ha, I see. What I saw on some user defined circuits is developers re-using circuits variable along the way, or doing some slice allocations in the Define. 
For example:
```
type Circuit struct {
A frontend.Variable
B []frontend.Variable
}

Define(...) {
circuit.A = api.Mul(circuit.A, circuit.A)
circuit.B = make([]frontend.Variable, ....)
}

```
which is definitely not thread safe --> 2 concurrent call to compile will not work.

Comment (ivokub) [reply circuitstats_test.go]: Do you think if trying to do it [here](https://github.com/ConsenSys/gnark/blob/master/test/assert.go#L371) would work? I.e. deep copy the circuit before first compilation and compare the circuit to the copied circuit after compilation?

Comment (gbotrel) [reply circuitstats_test.go]: I found deep-copying circuits from their interface (frontend.Circuit) to be more complex than expected :) In internal/utils I believe, there is a ShallowCopy method.

Comment (ivokub) [reply circuitstats_test.go]: I see. I created a new issue for it https://github.com/ConsenSys/gnark/issues/195.

**Review by gbotrel — APPROVED** [review]

**Pull request merged**